README.md
# *nodeset*'s readme

## Description

nodeset is a library and a program (`ns`) to manage nodesets written in Rust.
As of now one can iterate over nodes, count (`len()` method) them, display
in a folded way :

```rust
    use nodeset::node::Node;
    use std::process::exit;

    let node = match Node::new("r[1-10/2,15]esw[2-8]") {
        Ok(n) => n,
        Err(e) => {
            println!("Error: {}", e);
            exit(1);
        }
    };
    println!("Node string display : {}", "r[1-10/2,15]esw[2-8]");
    println!("Node normal display : {}", node);
    println!("Node debug display  : {:?}", node);
    println!("Node count          : {}", node.len());

    // use of the iterator
    for n in node {
        print!("{} ", n);
    }
```

It will print:
```terminal
Node string display : r[1-10/2,15]esw[2-8]
Node normal display : r[1-10/2,15]esw[2-8]
Node debug display  : Node { name: "r{}esw{}", sets: [RangeSet { set: [Range { start: 1, end: 10, step: 2, pad: 0, curr: 1 }, Range { start: 15, end: 15, step: 1, pad: 0, curr: 15 }], curr: 0 }, RangeSet { set: [Range { start: 2, end: 8, step: 1, pad: 0, curr: 2 }], curr: 0 }], values: [(0, 0), (0, 0)], first: true }
Node count          : 64
r1esw2 r1esw3 r1esw4 r1esw5 r1esw6 r1esw7 r1esw8 r3esw2 r3esw3 r3esw4 r3esw5 r3esw6 r3esw7 r3esw8 r5esw2 r5esw3 r5esw4 r5esw5 r5esw6 r5esw7 r5esw8 r7esw2 r7esw3 r7esw4 r7esw5 r7esw6 r7esw7 r7esw8 r9esw2 r9esw3 r9esw4 r9esw5 r9esw6 r9esw7 r9esw8 r15esw2 r15esw3 r15esw4 r15esw5 r15esw6 r15esw7 r15esw8
```

Alternatively you can use `node_to_vec_string()` function to directly expand node notation into a Vector of Strings :

```rust
use nodeset::node::{node_to_vec_string};

let v = node_to_vec_string("r[1-6/2]esw[1,3,5]-port[23-24]").unwrap();
assert_eq!(v, ["r1esw1-port23", "r1esw1-port24", "r1esw3-port23", "r1esw3-port24", "r1esw5-port23", "r1esw5-port24", "r3esw1-port23", "r3esw1-port24", "r3esw3-port23", "r3esw3-port24", "r3esw5-port23", "r3esw5-port24", "r5esw1-port23", "r5esw1-port24", "r5esw3-port23", "r5esw3-port24", "r5esw5-port23", "r5esw5-port24"]);
```

## ns

`ns` is a small and basic command line tool to manage nodeset(s). Use `-h` or `--help` to get commands that are usable.

See [INSTALL.md](https://gitlab.com/delhomme/nodeset/-/blob/master/INSTALL.md) file in the root path of the code for instruction on how to install this crate.

## Rust similar projects

* [nodeagg](https://crates.io/crates/nodeagg)
* [hostlist](https://crates.io/crates/hostlist)
* [hostlist-parser](https://crates.io/crates/hostlist-parser)


NEWS

# v0.4.1

 * Adds intersection method to Range, RangeSet and Node structures

# v0.4.0

 * Changes .amount() functions to .len() that are more idiomatic
 * Adds .is_empty() functions


src/lib.rs
/* -*- coding: utf8 -*-
 *
 *  lib.rs: Implements all structure logic to deal with NodeSets
 *
 *  (C) Copyright 2022 Olivier Delhomme
 *  e-mail : olivier.delhomme@free.fr
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 3, or (at your option)
 *  any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software Foundation,
 *  Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */
#![doc = include_str!("../README.md")]

/// module to manage node(s). Expanding for instance `node[1-4]` to `node1 node2 node3 node4`
pub mod node;

/// module to manage range such as `1-4` or `1` or even `30-0/4`
pub mod range;

/// module to manage a set of range called rangeset such as `1-4,8-14/2,50`
pub mod rangeset;


src/main.rs
/* -*- coding: utf8 -*-
 *
 *  lib.rs: Implements all structure logic to deal with nodesets
 *
 *  (C) Copyright 2022 Olivier Delhomme
 *  e-mail : olivier.delhomme@free.fr
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 3, or (at your option)
 *  any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software Foundation,
 *  Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

mod range;
mod rangeset;

use crate::range::Range;
use crate::rangeset::RangeSet;
use std::process::exit;

/// rack[10-49]node[1-25/2,78-89,101,1001].panel[0-30/4]
/// Between ',' a Range :
/// * 10-49
/// * 1-25/2,
/// * 78-89,
/// * 101,
/// * 1001
/// * 0-30/4
/// Between '[]' a Set
/// A global name 'rack{}node{}.panel{}' and a vector of sets.

fn print_range(range_str: &str) {
    let range = match Range::new(range_str) {
        Ok(r) => r,
        Err(e) => {
            println!("Error: {}", e);
            exit(1);
        }
    };
    println!("Range: {}", range_str);
    println!("Range: {}", range);
    println!("Range: {:?}", range);

    for i in range {
        println!("{}", i);
    }
}

fn print_rangeset(rangeset_str: &str) {
    let rangeset = match RangeSet::new(rangeset_str) {
        Ok(r) => r,
        Err(e) => {
            println!("Error: {}", e);
            exit(1);
        }
    };
    println!("RangeSet: {}", rangeset_str);
    println!("RangeSet: {}", rangeset);
    println!("RangeSet: {:?}", rangeset);

    for i in rangeset {
        println!("{}", i);
    }
}

fn main() {
    print_range("1-14/4");
    print_range("38-42");
    print_range("1");
    print_range("097-103");
    print_range("42-38");
    print_rangeset("1,3-5,89");
    print_rangeset("9-2,101,2-8/2");
    print_rangeset("9-2,101,2-8/2");
}


src/rangeset.rs
/* -*- coding: utf8 -*-
 *
 *  rangeset.rs: Implements all structure and logic to manage RangeSet
 *
 *  (C) Copyright 2022 Olivier Delhomme
 *  e-mail : olivier.delhomme@free.fr
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 3, or (at your option)
 *  any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software Foundation,
 *  Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

use crate::range::{Range, vec_u32_intersection};
use std::error::Error;
use std::fmt;
use std::fmt::Write;
use std::str::FromStr;

#[cfg(test)]
use std::process::exit; //used for testing

/// A RangeSet is a vector of Range.
/// Unlike Range a RangeSet may not be ordered. Specified order is
/// kept when generating numbers with the iterator.
/// * `set` is the vector of Range. It may be empty.
/// * `curr` is used remember the current index in the vector of Ranges
///      and is used to calculate next number in RangeSet iterator's
///      implementation.
///
/// RangeSet examples:
/// * "1,3-5,89"
/// * "9-2,101,2-8/2"
///
/// Example:
/// ```rust
/// use nodeset::rangeset::RangeSet;
/// let rangeset = RangeSet::new("22-28/2,29");
/// ```
#[derive(Debug)] /* Auto generates Debug trait */
pub struct RangeSet {
    set: Vec<Range>,
    curr: usize,
}

// This function needs a non empty sorted Vector of u32.
// It does fold every numbers in the vector into Ranges
// that are put in a vector. This vector contains at
// least one Range.
// pad will be used for all Range in the new Vector
fn fold_vec_u32_in_vec_range(v: Vec<u32>, pad: usize) -> Vec<Range> {

    let mut index = 0;
    let mut res: Vec<Range> = Vec::new();

    if v.len() == 1 {
        // only one value in the vector leads to only one Range with
        // start, end and curr at the same value and step to 1 (by convention)
        let range = Range::new_from_values(v[0] , v[0], 1,  pad, v[0]);
        res.push(range);
        res
    } else {
        // we know that we have at least two values
        // so index + 1 exists
        let mut step = v[index + 1] - v[index];
        let mut diff;
        let mut start = v[index];
        while index + 1 < v.len() {
            // println!("{index}");
            // If we have a third value ahead then begin the loop
            // until the end or until the difference between two
            // values changed
            while index + 2 < v.len() {
                diff = v[index + 2] - v[index + 1];
                if step != diff {
                    // When the difference between the next two values is
                    // not the same as the previous one, the range stops
                    // here (and pushed in the result vector) and a new
                    // one is started.
                    let end = v[index + 1];
                    let range = Range::new_from_values(start, end, step, pad, start);
                    res.push(range);
                    start = v[index + 2];
                    if index + 3 < v.len() {
                        step = v[index + 3] - v[index + 2];
                    } else {
                        step = 1;
                    }
                    break;
                } else {
                    index += 1;
                }
            }
            index += 1;
        }

        let end = v[index];
        let range = Range::new_from_values(start, end, step, pad, start);
        res.push(range);
        res
    }
}

impl RangeSet {
    /// True when we only have one member and not a set ie: node003
    pub fn is_alone(&self) -> bool {
        self.set.len() == 1 && self.set[0].start_is_end() && self.set[0].step_is_one()
    }

    pub fn reset(&mut self) {
        self.curr = 0;
        for i in 0..self.set.len() {
            self.set[i].reset()
        }
    }

    pub fn get_current(&self) -> (u32, usize) {
        let index = self.curr;
        let pad = self.set[index].get_pad();

        (self.set[index].get_current(), pad)
    }

    /// Counts the number of elements in the rangeset
    pub fn len(&self) -> u32 {
        if self.set.is_empty() {
            0
        } else {
            let mut total = 0;
            for r in self.set.iter() {
                total += r.len();
            }
            total
        }
    }

    /// Tells whether a RangeSet is empty or not.
    pub fn is_empty(&self) -> bool {
        self.set.is_empty()
    }

    /// Intersection of self RangeSet with other RangeSet :
    ///  `1,3-5,89` and `9-2,101,2-8/2`
    pub fn intersection(&self, other: &Self) -> Option<RangeSet> {

        // special cases where self or other is empty
        if self.is_empty() {
            return Some(RangeSet{set: other.set.clone(), curr: other.curr});
        } else if other.is_empty() {
            return Some(RangeSet{set: self.set.clone(), curr: self.curr});
        }
        // here self and other are not empty so we get at least
        // 2 vectors.

        let mut first: Vec<u32> = Vec::new();
        let mut second: Vec<u32> = Vec::new();
        let mut pad: usize = 0;

        for r in &self.set {
            pad = pad.max(r.get_pad());
            let mut v = r.generate_vec_u32();
            first.append(& mut v);
        }
        for r in &other.set {
            pad = pad.max(r.get_pad());
            let mut v = r.generate_vec_u32();
            second.append(& mut v);
        }

        if let Some(inter) = vec_u32_intersection(first, second) {
            //println!("{:?}", inter);
            let range_vec = fold_vec_u32_in_vec_range(inter, pad);
            //println!("{:?}", range_vec);
            Some(RangeSet {set: range_vec, curr: 0})
        } else {
            None
        }
    }

    pub fn get_next(&mut self) -> Option<(u32, usize)> {
        let index = self.curr;
        let mut pad = self.set[index].get_pad();

        let next = match self.set[index].get_next() {
            Some(number) => number, // gives next number in Range range.
            None => {
                /* This tells us that range Range is finished : need to iter over next range. */
                if index + 1 < self.set.len() {
                    /* There is another Range in the vector */
                    self.curr = index + 1;
                    pad = self.set[self.curr].get_pad();
                    match self.set[self.curr].get_next() {
                        Some(number) => number,
                        None => return None,
                    }
                } else {
                    /* There is no other Range in the vector */
                    return None;
                }
            }
        };
        Some((next, pad))
    }

    /// "[1-5/2]" or "[1,3-5,89]" or "[9-15/3,4,9-2]"
    pub fn new(strange: &str) -> Result<RangeSet, Box<dyn Error>> {
        let mut set: Vec<Range> = Vec::new();
        let rangeset: Vec<&str> = strange.split(',').collect();
        let curr = 0;

        for rs in rangeset {
            let range = match Range::new(rs) {
                Ok(r) => r,
                Err(e) => return Err(e),
            };
            set.push(range);
        }
        Ok(RangeSet {
            set,
            curr,
        })
    }

    pub fn empty() -> RangeSet {
        let set: Vec<Range> = Vec::new();
        let curr = 0;

        RangeSet {
            set,
            curr,
        }
    }
}

/// RangeSet iterator returns an already padded String as Range does.
impl Iterator for RangeSet {
    type Item = String;

    fn next(&mut self) -> Option<Self::Item> {
        let (next_u32, pad) = match self.get_next() {
            Some(v) => v,
            None => return None,
        };

        let next = format!("{:0pad$}", next_u32);
        Some(next)
    }
}

/// FromStr trait lets you write: `let a_rangeset: RangeSet = "01-10/2,15-30/3".parse().unwrap();`
impl FromStr for RangeSet {
    type Err = Box<dyn Error>;

    fn from_str(strange: &str) -> Result<Self, Self::Err> {
        RangeSet::new(strange)
    }
}

/// PartialEq trait for RangeSet to know if a rangeSet is equal or not
/// to another rangeSet. curr (Iterator's position) is not taken into
/// account. A RangeSet is equal to another one when all ranges are
/// equal to each others in the same order (order matters).
impl PartialEq for RangeSet {
    fn eq(&self, other: &Self) -> bool {
        let mut ok: bool = true;
        if self.set.len() == other.set.len() {
            for i in 0..self.set.len() {
                ok = ok && self.set[i] == other.set[i]
            }
            ok
        } else {
            false
        }
    }
}

/// Display trait for RangeSet. It will display the RangeSet in a folded way
impl fmt::Display for RangeSet {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let mut to_display = String::new();
        let len = self.set.len();

        for (i, range) in self.set.iter().enumerate() {
            if i == len - 1 {
                write!(&mut to_display, "{}", range).unwrap();
            } else {
                write!(&mut to_display, "{},", range).unwrap();
            }
        }

        write!(f, "{}", to_display)
    }
}

/*********************************** Tests ***********************************/

#[cfg(test)] /* Helper function for testing */
fn get_rangeset_values_from_str(rangeset_str: &str) -> Vec<String> {
    let rangeset = match RangeSet::new(rangeset_str) {
        Ok(r) => r,
        Err(e) => {
            println!("Error: {}", e);
            exit(1);
        }
    };
    let mut v: Vec<String> = Vec::new();
    for r in rangeset {
        v.push(r);
    }
    v
}

#[test]
fn testing_creating_rangeset() {
    let rangeset = RangeSet::new("1-10").unwrap();
    let range = Range::new("1-10").unwrap();
    assert_eq!(
        rangeset,
        RangeSet {
            set: vec![range],
            curr: 0
        }
    );

    let rangeset = RangeSet::new("1,2,3-10").unwrap();
    let range_a = Range::new("1").unwrap();
    let range_b = Range::new("2").unwrap();
    let range_c = Range::new("3-10").unwrap();
    assert_eq!(
        rangeset,
        RangeSet {
            set: vec![range_a, range_b, range_c],
            curr: 0
        }
    );

    let rangeset = RangeSet::new("1,2,3-10").unwrap();
    let range_a = Range::new("1").unwrap();
    let range_b = Range::new("2").unwrap();
    let range_c = Range::new("3-10").unwrap();
    assert_ne!(
        rangeset,
        RangeSet {
            set: vec![range_b, range_a, range_c],
            curr: 0
        }
    );
}

#[test]
fn testing_rangeset_values() {
    let value = get_rangeset_values_from_str("1,3-5,89");
    assert_eq!(value, vec!["1", "3", "4", "5", "89"]);

    let value = get_rangeset_values_from_str("9-2,101,2-8/2");
    assert_eq!(value, vec!["9", "8", "7", "6", "5", "4", "3", "2", "101", "2", "4", "6", "8"]);

    let value = get_rangeset_values_from_str("10-01/2,32-72/4");
    assert_eq!(value, vec!["10", "08", "06", "04", "02", "32", "36", "40", "44", "48", "52", "56", "60", "64", "68", "72"]);

    let value = get_rangeset_values_from_str("01-10,7-12/2");
    assert_eq!(value, vec!["01", "02", "03", "04", "05", "06", "07", "08", "09", "10", "7", "9", "11"]);
}

#[test]
fn testing_rangeset_intersection() {
    let rs_a:RangeSet = "1,3-5,89".parse().unwrap();
    // "1", "3", "4", "5", "89"
    let rs_b:RangeSet = "9-2,101,2-8/2,89".parse().unwrap();
    // "9", "8", "7", "6", "5", "4", "3", "2", "101", "2", "4", "6", "8", "89"

    let inter = rs_a.intersection(&rs_b);
    let range_a = Range::new("3-5").unwrap();
    let range_b = Range::new("89").unwrap();
    println!("{:?}", inter);
    assert_eq!(
        inter,
        Some(RangeSet { set : vec![range_a, range_b], curr: 0 }));


    let rs_a:RangeSet = "10-01/2,32-72/4".parse().unwrap();
    // "10", "08", "06", "04", "02", "32", "36", "40", "44", "48", "52", "56", "60", "64", "68", "72"
    let rs_b:RangeSet = "01-10,7-12/2,50-60/2".parse().unwrap();
    // "01", "02", "03", "04", "05", "06", "07", "08", "09", "10", "7", "9", "11"

    let inter = rs_a.intersection(&rs_b);
    let range_a = Range::new("02-10/2").unwrap();
    let range_b = Range::new("52-60/4").unwrap();
    println!("{:?}", inter);
    assert_eq!(
        inter,
        Some(RangeSet { set : vec![range_a, range_b], curr: 0 }));
}


src/node.rs
/* -*- coding: utf8 -*-
 *
 *  node.rs: Implements all structure and logic to manage Node
 *
 *  (C) Copyright 2022 Olivier Delhomme
 *  e-mail : olivier.delhomme@free.fr
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 3, or (at your option)
 *  any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software Foundation,
 *  Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

use crate::rangeset::RangeSet;
use lazy_static::lazy_static;
use regex::Regex;
use std::error::Error;
use std::fmt;
use std::str::FromStr;

#[cfg(test)]
use std::process::exit;

/// A Node is a name that may contain multiple RangeSets and
/// that defnines a machine name. For instance `node[1-14]` is
/// a valid Node defining 14 nodes name from node1 to node14.
/// Let's say that within these nodes we want to reference each
/// core (32 per cpu) of each cpu (2 per node) we could do this
/// writing `node[1-14]-cpu[1-2]-core[1-32]`.
///
/// ```rust
///
/// use nodeset::node::Node;
/// use std::process::exit;
/// let node = match Node::new("r1esw[2-6]") {
///     Ok(n) => n,
///     Err(e) => {
///         println!("Error: {}", e);
///         exit(1);
///     }
/// };
/// let v: Vec<_> = node.into_iter().map(|x| x).collect();
/// assert_eq!(v, ["r1esw2", "r1esw3", "r1esw4", "r1esw5", "r1esw6"]);
/// ```
/// Note : to transform a node into a vector of Strings you may
///        prefer to use `node_to_vec_string()` function.

/*
 * Structure used to keep Node definition
 * * name is the name of the node where everything between brackets
 *        (and the brackets themselves) is replaced by '{}'.
 * * sets is a vector of rangesets: one rangeset per brackets found.
 * * values is used to compute the iterator (and get_next) method
 *          and is a tuple (index, pad) corresponding to the RangeSet
 *          at the same index in the vector
 * * first is also used to compute the iterator and is true until
 *         the first time we pass into the iterator.
 */
#[derive(Debug)]
pub struct Node {
    name: String,
    sets: Vec<RangeSet>,
    values: Vec<(u32, usize)>,
    first: bool,
}

#[derive(Debug)]
pub enum NodeErrorType {
    Regular(ErrorKind),
}

#[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
pub enum ErrorKind {
    RegexNoMatch,
    RangeSetCreation,
}

impl ErrorKind {
    fn as_str(&self) -> &str {
        match *self {
            ErrorKind::RegexNoMatch => "no match found in string",
            ErrorKind::RangeSetCreation => "unable to create rangeset",
        }
    }
}

impl fmt::Display for NodeErrorType {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match *self {
            NodeErrorType::Regular(ref err) => write!(f, "Error: {:?}", err),
        }
    }
}

impl Error for NodeErrorType {
    fn description(&self) -> &str {
        match *self {
            NodeErrorType::Regular(ref err) => err.as_str(),
        }
    }
}

/// Transforms a nodeset (String) into a vector of nodes (String)
/// by expanding the created Node structure. This method can be
/// very expensive on memory consumption depending on the number
/// of nodes once that are expanded.
/// ```rust
/// use nodeset::node::{node_to_vec_string};
///
/// let v = node_to_vec_string("r1esw[2-6]").unwrap();
/// assert_eq!(v, ["r1esw2", "r1esw3", "r1esw4", "r1esw5", "r1esw6"]);
/// ```
pub fn node_to_vec_string(node_str: &str) -> Result<Vec<String>, Box<dyn Error>> {
    let node = match Node::new(node_str) {
        Ok(n) => n,
        Err(e) => return Err(Box::new(e)),
    };
    let v: Vec<String> = node.into_iter().collect();
    Ok(v)
}

/* This regular expression is used to capture each rangeset in a string defining a Node */
lazy_static! {
    static ref RE: Regex = Regex::new(r"\[([\d,\-/]+)\]|([\d]+)").unwrap();
}

impl Node {
    /// Counts the number of elements in Node's definition.
    pub fn len(&self) -> u32 {
        if self.sets.is_empty() {
            if self.name.is_empty() {
                0
            } else {
                1
            }
        } else {
            let mut total = 1;
            for r in self.sets.iter() {
                total *= r.len();
            }
            total
        }
    }
    /// Tells whether a Node is empty or not.
    pub fn is_empty(&self) -> bool {
        self.sets.is_empty() && self.name.is_empty()
    }

    /// Intersection of self Node with an other Node :
    ///  `node[1,3-5,89]-cpu[2-4]` and `node[9-2,89,101,2-8/2]-cpu[1-3]`
    ///  -> `node[3-5,89]-cpu[2-3]`
    pub fn intersection(&self, other: &Self) -> Option<Node> {

        let mut ns_sets: Vec<RangeSet> = Vec::new();
        let mut values: Vec<(u32, usize)> = Vec::new();

        if self.name != other.name {
            None
        } else {
            for (i, rs_a) in self.sets.iter().enumerate() {
                let rs_b: &RangeSet = &other.sets[i];
                if let Some(inter) = rs_a.intersection(rs_b) {
                    ns_sets.push(inter);
                    values.push((0, 0));
                } else {
                    return None;
                }
            }
            Some(Node { name: self.name.to_string(), sets: ns_sets, values, first: false })
        }
    }

    /* Captures with regex all possible (and non overlapping) rangeset in the node name
     * for instance rack[1-8]-node[1-42] should return 1-8 and 1-42 as rangeset
     * It will capture mixed types of rangesets ie: rack1-node[1-42]-cpu2
     */
    fn capture_with_regex(nodename: &str) -> Result<(String, Vec<String>), NodeErrorType> {
        let mut rangesets: Vec<String> = Vec::new();
        let mut name = nodename.to_string();
        for capture in RE.captures_iter(nodename) {
            // println!("capture: {:?}", capture);
            match capture.get(1) {
                Some(text) => rangesets.push(text.as_str().to_string()),
                None => {
                    if let Some(text) = capture.get(2) {
                        rangesets.push(text.as_str().to_string())
                    };
                }
            };
        }
        if !rangesets.is_empty() {
            name = RE.replace_all(nodename, "{}").to_string();
        }
        // println!("name: {}", name);

        Ok((name, rangesets))
    }

    /// Node examples: "node[1-5/2]" or "rack[1,3-5,89]" or "cpu[1-2]core[1-64]" or "node01"
    pub fn new(str: &str) -> Result<Node, NodeErrorType> {
        let (name, rangesets) = Node::capture_with_regex(str)?;
        let mut sets: Vec<RangeSet> = Vec::new();
        let mut values: Vec<(u32, usize)> = Vec::new();
        for set in rangesets {
            let rangeset = match RangeSet::new(&set) {
                Ok(r) => r,
                Err(_) => return Err(NodeErrorType::Regular(ErrorKind::RangeSetCreation)),
            };
            sets.push(rangeset);
            values.push((0, 0));
        }

        Ok(Node {
            name,
            sets,
            values,
            first: true,
        })
    }

    fn make_node_string(&self) -> String {
        let mut nodestr: &str = self.name.as_str();
        let mut replaced;

        for i in 0..self.sets.len() {
            let (current, pad) = self.values[i];
            replaced = nodestr.replacen("{}", format!("{:0pad$}", current).as_str(), 1);
            nodestr = replaced.as_str();
        }

        nodestr.to_string()
    }

    fn get_next(&mut self) -> Option<(u32, usize)> {
        for i in (0..self.sets.len()).rev() {
            //println!("{}: {:?}", i, self.sets[i]);
            match self.sets[i].get_next() {
                Some(v) => {
                    //println!("{}: {:?} - {}", i, self.sets[i], v);
                    self.values[i] = v;
                    return Some(v);
                }
                None => {
                    self.sets[i].reset();
                    self.values[i] = self.sets[i].get_current();
                    self.sets[i].get_next();
                }
            };
        }
        None
    }
}

/// Iterator implementation for Node to allow one to use `for n in node {...}` construction.
impl Iterator for Node {
    type Item = String;

    fn next(&mut self) -> Option<Self::Item> {
        if self.sets.is_empty() {
            if self.first {
                self.first = false;
                Some(self.name.to_string())
            } else {
                None
            }
        } else {
            if self.first {
                self.first = false;
                for i in 0..self.sets.len() {
                    self.values[i] = match self.sets[i].get_next() {
                        Some(v) => v,
                        None => self.sets[i].get_current(),
                    };
                }
                return Some(self.make_node_string());
            }

            match self.get_next() {
                Some(_) => Some(self.make_node_string()),
                None => None,
            }
        }
    }
}

/// FromStr trait lets you write: `let a_node: Node = "node[1-6]-socket[1-2]-core[1-64]".parse().unwrap();`
impl FromStr for Node {
    type Err = NodeErrorType;

    fn from_str(node_str: &str) -> Result<Self, Self::Err> {
        Node::new(node_str)
    }
}

/// PartialEq trait for Node to know if a Node is equal or not
/// to another Node. curr (Iterator's position) is not taken into
/// account. Nodes are equal if name is equal and all RangeSets
/// are equal in the same order (order matters).
impl PartialEq for Node {
    fn eq(&self, other: &Self) -> bool {
        if self.name != other.name {
            return false;
        }

        let mut ok: bool = true;
        if self.sets.len() == other.sets.len() {
            for i in 0..self.sets.len() {
                ok = ok && self.sets[i] == other.sets[i]
            }
            ok
        } else {
            false
        }
    }
}

/// Display trait for Node. It will display the node in a folded way (node[1-9/2,98])
impl fmt::Display for Node {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let mut nodestr: &str = self.name.as_str();
        let mut replaced;
        for set in &self.sets {
            if set.is_alone() {
                replaced = nodestr.replacen("{}", format!("{}", set).as_str(), 1)
            } else {
                replaced = nodestr.replacen("{}", format!("[{}]", set).as_str(), 1)
            };
            nodestr = replaced.as_str();
        }
        write!(f, "{}", nodestr)
    }
}

/*********************************** Tests ***********************************/

#[cfg(test)] /* Helper function for testing */
fn get_node_values_from_str(node_str: &str) -> Vec<String> {
    let node = match Node::new(node_str) {
        Ok(r) => r,
        Err(e) => {
            println!("Error: {}", e);
            exit(1);
        }
    };
    let mut v: Vec<String> = Vec::new();
    for n in node {
        v.push(n);
    }
    v
}

#[test]
fn testing_creating_node() {
    let node: Node = "node[1-10]".parse().unwrap();
    let rangeset = RangeSet::new("1-10").unwrap();
    assert_eq!(
        node,
        Node {
            name: "node{}".to_string(),
            sets: vec![rangeset],
            values: vec![(0, 0)],
            first: false
        }
    );

    let node: Node = "node[1-10]-cpu[1,2]-core[1-32,34-64]".parse().unwrap();
    let rangeset_a = RangeSet::new("1-10").unwrap();
    let rangeset_b = RangeSet::new("1,2").unwrap();
    let rangeset_c = RangeSet::new("1-32,34-64").unwrap();
    assert_eq!(
        node,
        Node {
            name: "node{}-cpu{}-core{}".to_string(),
            sets: vec![rangeset_a, rangeset_b, rangeset_c],
            values: vec![(0, 0), (0, 0), (0, 0)],
            first: false
        }
    );
    let node: Node = "node[1-10]-cpu[1-2]-core[1-32,34-64]".parse().unwrap();
    let rangeset_a = RangeSet::new("1-10").unwrap();
    let rangeset_b = RangeSet::new("1-2").unwrap();
    let rangeset_c = RangeSet::new("1-32,34-64").unwrap();
    assert_ne!(
        node,
        Node {
            name: "node{}-cpu{}-core{}".to_string(),
            sets: vec![rangeset_c, rangeset_b, rangeset_a],
            values: vec![(0, 0), (0, 0), (0, 0)],
            first: false
        }
    );
}

#[test]
fn testing_nodes_values() {
    let value = get_node_values_from_str("r[1-6]esw[1-3]");
    assert_eq!(
        value,
        vec!["r1esw1", "r1esw2", "r1esw3", "r2esw1", "r2esw2", "r2esw3", "r3esw1", "r3esw2", "r3esw3", "r4esw1", "r4esw2", "r4esw3", "r5esw1", "r5esw2", "r5esw3", "r6esw1", "r6esw2", "r6esw3"]
    );

    let value = get_node_values_from_str("node[01-10,7-12/2]");
    assert_eq!(value, vec!["node01", "node02", "node03", "node04", "node05", "node06", "node07", "node08", "node09", "node10", "node7", "node9", "node11"]);

    let value = get_node_values_from_str("node001");
    assert_eq!(value, vec!["node001"]);

    let value = get_node_values_from_str("node[1]");
    assert_eq!(value, vec!["node1"]);

    let value = get_node_values_from_str("r1esw[2-6]");
    assert_eq!(value, vec!["r1esw2", "r1esw3", "r1esw4", "r1esw5", "r1esw6"]);

    let value = get_node_values_from_str("toto");
    assert_eq!(value, vec!["toto"]);

    let value = get_node_values_from_str("r[1-7/2,15]esw[2-4]");
    assert_eq!(value, vec!["r1esw2", "r1esw3", "r1esw4", "r3esw2", "r3esw3", "r3esw4", "r5esw2", "r5esw3", "r5esw4", "r7esw2", "r7esw3", "r7esw4", "r15esw2", "r15esw3", "r15esw4"]);

    let value = get_node_values_from_str("rack1-node[1-3]-cpu2");
    assert_eq!(value, vec!["rack1-node1-cpu2", "rack1-node2-cpu2", "rack1-node3-cpu2"]);

    let value = get_node_values_from_str("rack[1-2]-node[1-2]-cpu[1-2]");
    assert_eq!(value, vec!["rack1-node1-cpu1", "rack1-node1-cpu2", "rack1-node2-cpu1", "rack1-node2-cpu2", "rack2-node1-cpu1", "rack2-node1-cpu2", "rack2-node2-cpu1", "rack2-node2-cpu2"]);
}

#[test]
fn testing_node_intersection() {
    let ns_a:Node = "node[1,3-5,89]-cpu[2-4,85-90]".parse().unwrap();
    let ns_b:Node = "node[9-2,89,101,2-8/2]-cpu[1-3,86-92/2]".parse().unwrap();
    // -> node[3-5,89]-cpu[2-3,86-90/2]

    let inter = ns_a.intersection(&ns_b);
    let rs_a = RangeSet::new("3-5,89").unwrap();
    let rs_b = RangeSet::new("2-3,86-90/2").unwrap();
    println!("{:?}", inter);
    assert_eq!(
        inter,
        Some(Node {name: "node{}-cpu{}".to_string(), sets: vec![rs_a, rs_b], values: vec![(0, 0), (0,0)], first: false}));

    let ns_a:Node = "node[1,3-5,89]-cpu[2-4]".parse().unwrap();
    let ns_b:Node = "node[9-2,89,101,2-8/2]-cpu[16-32]".parse().unwrap();
    // -> no intersection !

    let inter = ns_a.intersection(&ns_b);
    println!("{:?}", inter);
    assert_eq!(
        inter,
        None);
}
